aspnetcore/signalr/javascript-client.md

```diff
@@ -0,0 +1,105 @@
+
+Reference the SignalR JavaScript client in the `<script>` element.
+
+```html
+<script src="~/wwwroot/lib/signalr/signalr.js"></script>
+```
+
+## Connect to a hub
+
+The following code creates a connection. The hub's name is case insensitive.
+
+```javascript
+const connection = new signalR.HubConnection('/chathub');
+```
+
+While the preceding code sample creates a connection to a hub named `ChatHub`, it doesn't start it. To start the connection, call the `connection.start` method.
+
+[!code-javascript[Call hub methods](javascript-client/sample/chat.js?range=9)]
+
+### Cross-origin connections
+
+Typically, browsers load connections from the same domain as the requested page. However, there are occasions when a connection to another domain is required.
+
+To prevent a malicious site from reading sensitive data from another site, [cross-origin connections](xref:security/cors) are disabled by default. To allow a cross-origin request, enable it in the `Startup` class.
+
+[!code-csharp[Cross-origin connections](javascript-client/sample/startup.cs?highlight=22-29,39-40)]
+
+## Call hub methods from client
+
+JavaScript clients call public methods on hubs by issuing `connection.invoke`. The `invoke` method accepts two arguments:
+
+* The name of the hub method. In the following example, the hub name is `SendMessage`.
+* Any arguments defined in the hub method. In the following example, the argument name is `message`.
+
+[!code-javascript[Call hub methods](javascript-client/sample/chat.js?range=12)]
+
+## Call client methods from hub
+
+To receive messages from the hub, define a method in the client's `connection.on` method. The `connection.on` method requires the following information:
+
+* The name of the JavaScript client method. In the following example, the method name is `ReceiveMessage`.
+* Arguments the hub passes to the method. In the following example, the argument value is `message`.
+
+[!code-javascript[Receive calls from hub](javascript-client/sample/chat.js?range=4-7)]
+
+The preceding code in `connection.on` runs when  server-side code calls it using the `SendAsync` method.
+
+[!code-javascript[Call client-side](javascript-client/sample/chathub.cs?range=8-11)]
+
+SignalR determines which client method to call by matching the method name and arguments defined in `SendAsync` and `connection.on`.
+
+> [!NOTE]
+> As a best practice, call `connection.start` after `connection.on` so your handlers are registered before any messages are received.
+
+## Error handling and logging
+
+Chain a `catch` method to the end of the `connection.start` method to handle client-side errors. Use `console.error` to output errors to the browser's console.
+
+[!code-javascript[Error handling](javascript-client/sample/chat.js?range=9)]
+
+Setup client-side log tracing by passing a logger and type of event to log when the connection is made. Available log levels are as follows:
+
+* `signalR.LogLevel.Information` : Status messages without errors.
```

Comment:
> Status messages without errors.

Not quite accurate, if you specify a log level, we'll log messages at that level **or above**. The highest level is `Error`, and the lowest is `Trace` (which is much more detailed and not generally useful for users). So if you set the log level to `Information`, you'll get `Information`, `Warning` and `Error` messages logged. If you set it to `Warning`, you'll just get `Warning` and `Error`, if you set it to `Trace`, you get every dang thing ;).